test/cli/test_session.rb

```diff
@@ -0,0 +1,155 @@
+# frozen_string_literal: true
+
+require "helpers"
+
+# tests cli command: session
+class CLISessionTest < Minitest::Test
+  include Aruba::Api
+  include Helpers
+
+  def setup
+    setup_aruba
+  end
+
+  def test_session
+    session_1 = "session_one"
+    session_2 = "session_two"
+    session_1_label = "session-one-unstructured"
+    session_2_label = "session-two-unstructured"
+    session_1_text = "session one unstructured data"
+    session_2_text = "session two unstructured data"
+    session_1_output = "text:#{session_1_text}"
+    session_2_output = "text:#{session_2_text}"
+    error_pattern = /No such record/
+
+    # create a session:
+    run_command_and_stop "kbsecret session new -r test #{session_1}"
+
+    # create record in session:
+    run_command "kbsecret new unstructured -s #{session_1} -x #{session_1_label}" do |cmd|
+      cmd.stdin.puts session_1_text
+      cmd.stdin.close
+      cmd.wait
+    end
+
+    # retrieve record from session:
+    run_command "kbsecret dump-fields -s #{session_1} -x #{session_1_label}" do |cmd|
+      cmd.wait
+      assert_equal session_1_output, cmd.output.chomp
+    end
+
+    # create another session:
+    run_command_and_stop "kbsecret session new -r test #{session_2}"
+
+    # create record in second session:
+    run_command "kbsecret new unstructured -s #{session_2} -x #{session_2_label}" do |cmd|
+      cmd.stdin.puts session_2_text
+      cmd.stdin.close
+      cmd.wait
+    end
+
+    # retrieve record from second session:
+    run_command "kbsecret dump-fields -s #{session_2} -x #{session_2_label}" do |cmd|
+      cmd.wait
+      assert_equal session_2_output, cmd.output.chomp
+    end
+
+    # FIXME: this test actually fails.
```

Comment:
Yes, this was my intention. The "same root" refers to the same directory, so two sessions with the same root are really the same session known by two different labels. The reason for this session-root divide was originally to facilitate session sharing, but that's now been accomplished in other ways.

Thanks for pointing this out. I'm going to look at ways to fix this this week.

Reply:
On reflection, this behavior probably isn't going to change -- I may make it so that the user doesn't have any control over the session root directory's name (no more `-r` for `session new`), but existing sessions will continue to work as they currently do.

The use case this handles is as follows:

1. Bob has one session, labeled `foo`.
2. Alice has two sessions, labeled `bar`, and `baz`.
3. Bob creates a new session labeled `bar` and shares it with Alice.
4. Alice already has a `bar`-labeled session, and she doesn't want to overwrite or rename it. Therefore, she configures Bob's `bar` as `bar-bob`. As a result, the same session is known by two different labels (albeit to two different users).